### PR TITLE
only set featuregates string if non empty

### PR DESCRIPTION
**Pull request by BenTheElder**
see this thread: https://github.com/kubernetes-sigs/kind/pull/1263/files#r368245568

untested

### CODE REVIEW

**Comment by BenTheElder** [comment]
cc @liggitt @aojea

File: pkg/cluster/internal/kubeadm/config.go [hunk 1]
```diff
@@ -415,8 +420,10 @@ controlPlaneEndpoint: "{{ .ControlPlaneEndpoint }}"
 # to the cluster after rewriting the kubeconfig to point to localhost
 apiServer:
   certSANs: [localhost, "{{.APIServerAddress}}"]
+  {{ if .FeatureGates -}}
   extraArgs:
     "feature-gates": "{{ .FeatureGatesString }}"
+  {{- end }}
 controllerManager:
   extraArgs:
     "feature-gates": "{{ .FeatureGatesString }}"
```
Comment (liggitt) [inline pkg/cluster/internal/kubeadm/config.go]: needs wrapping here?

Comment (BenTheElder) [reply pkg/cluster/internal/kubeadm/config.go]: thanks, done

**Comment by aojea** [comment]
having second thoughts about this, because it only solves the problem when no feature-gate is configured that may be confusing for the users when they try to use additional feature gates and only some are working.

At the end of the day, we are only trying to solve the dual-stack problem that has 2 different features gates, one in kubeadm that configures the corresponding features gates of the components and does validation on the input parameters.

I think that we should go with Ben suggestion of implementing kubeadm DualStack feature gate in kind (validation and population of components featureGates) and avoid this change.

**Comment by BenTheElder** [comment]
> having second thoughts about this, because it only solves the problem when no feature-gate is configured that may be confusing for the users when they try to use additional feature gates and only some are working.

This is a debateably useful config change that just avoids explicitly setting empty strings all over the place, which is sort of noise. On the other hand it adds noise to the config templates...

> At the end of the day, we are only trying to solve the dual-stack problem that has 2 different features gates, one in kubeadm that configures the corresponding features gates of the components and does validation on the input parameters.

I am not trying to solve that here. The solution to that is to have kind internally handle dualstack properly, kind's networking config should drive feature gates etc, which is totally unrelated to this PR.

> I think that we should go with Ben suggestion of implementing kubeadm DualStack feature gate in kind (validation and population of components featureGates) and avoid this change.

Yeah ... To clarify though: 
- We already enable _kubernetes_ feature gate(s) in kind in kubernetes 1.11
- kubeadm feature gates can influence kubernetes feature gates but we should not rely on that if we want to allow users to enable extra feature gates... Regardless of anything kind did special already, this is expected in kubeadm, if you want feature gate foo AND kubeadm's feature gate bar that enables kubernetes feature gate baz then you must set foo and baz as kubernetes feature gates yourself. it won't mutate an explicit setting of foo to append baz.
- kind user supplied feature gates map[name]=enabled should be applied on top of the kind generate feature gates before passing to kubeadm

**Comment by aojea** [comment]
very convincing 
/hold cancel

File: pkg/cluster/internal/kubeadm/config.go [hunk 2]
```diff
@@ -311,15 +312,19 @@ apiServer:
...
```
Comment (aojea) [inline pkg/cluster/internal/kubeadm/config.go]: this should be added to the `if`

File: pkg/cluster/internal/kubeadm/config.go [hunk 3]
```diff
@@ -311,15 +312,19 @@ apiServer:
...
 controllerManager:
   extraArgs:
+{{ if .FeatureGates -}}
```
Comment (aojea) [inline pkg/cluster/internal/kubeadm/config.go]: move up a few lines so it includes the apiserver feature-gates

**Comment by liggitt** [comment]
lgtm, isn't this still needed?

**Comment by BenTheElder** [comment]
reviving...

File: pkg/cluster/internal/kubeadm/config.go (hunk 3 above)
Comment (BenTheElder) [reply pkg/cluster/internal/kubeadm/config.go]: sigh, this part was correct, moving it up broke the other extraArgs.

**Comment by BenTheElder** [comment]
https://github.com/kubernetes-sigs/kind/pull/1274#discussion_r386720341 ... fixing

**Comment by BenTheElder** [comment]
flake
> 

[sig-node] ConfigMap should be consumable via the environment [NodeConformance] [Conformance] expand_less
--


https://prow.k8s.io/view/gcs/kubernetes-jenkins/pr-logs/pull/sigs.k8s.io_kind/1274/pull-kind-conformance-parallel-1-15/1234627209726005251/
/retest

**Comment by BenTheElder** [comment]
tests green

**Comment by BenTheElder** [comment]
Flakes
/retest

**Pull request merged**